src/api/server.rs

```diff
@@ -0,0 +1,123 @@
+use super::{handler, schema::build_schema};
+
+use async_graphql::{
+    http::{playground_source, GraphQLPlaygroundConfig},
+    QueryBuilder,
+};
+use async_graphql_warp::{graphql_subscription, GQLResponse};
+use std::convert::Infallible;
+use std::net::SocketAddr;
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::{Receiver, Sender};
+use warp::filters::BoxedFilter;
+use warp::{http::Response, Filter, Reply};
+
+pub struct Server {
+    /// Address for the API server to bind on
+    address: SocketAddr,
+
+    /// Transmission channel to trigger closure of a running API server
+    trigger_cancel: Sender<()>,
+
+    /// Receiver signal to cancel running API server
+    cancel_signal: Option<Receiver<()>>,
+}
+
+impl Server {
+    /// Returns a new API Server
+    pub fn new(address: SocketAddr) -> Server {
+        let (trigger_cancel, cancel_signal) = oneshot::channel::<()>();
+
+        Server {
+            address,
+            trigger_cancel,
+            cancel_signal: Some(cancel_signal),
+        }
+    }
+
+    /// String representation of the bound IP address
+    pub fn ip(&self) -> String {
+        self.address.ip().to_string()
+    }
+
+    /// String representation of the bound port
+    pub fn port(&self) -> String {
+        self.address.port().to_string()
+    }
+
+    /// Stops the running API server
+    pub fn stop(self) {
+        let _ = self.trigger_cancel.send(());
+    }
+
+    /// Run the API server
+    pub async fn run(mut self) -> Self {
+        let rx = self
+            .cancel_signal
+            .take()
+            .expect("Run can only be called once");
+
+        let (_, server) =
+            warp::serve(make_routes()).bind_with_graceful_shutdown(self.address, async move {
+                let _ = rx.await;
+            });
+
+        tokio::spawn(server);
+
+        self
+    }
+}
+
+/// Builds Warp routes, to be served
+fn make_routes() -> BoxedFilter<(impl Reply,)> {
+    // Health route
+    let health_route = warp::path("health").and_then(handler::health);
+
+    // Build the GraphQL schema
+    let schema = build_schema().finish();
+
+    // GraphQL POST handler
+    let graphql_post = async_graphql_warp::graphql(schema.clone()).and_then(
+        |(schema, builder): (_, QueryBuilder)| async move {
+            let resp = builder.execute(&schema).await;
+            Ok::<_, Infallible>(GQLResponse::from(resp))
+        },
+    );
+
+    // GraphQL playground. Also sets up /graphql as an endpoint for both queries + subscriptions
+    let graphql_playground = warp::path("playground").map(|| {
```

Comment:
I was thinking it might be nice to have this configurable, maybe...

```
[api]
  playground = true
```

Thinking it might be useful to have something to query directly, outside of the main UI.

What do you think?

Reply:
It seems like it could bloat the binary, but I'm not sure how much data is embedded to support the playground (although this implies we'd need a way to remove that data from being embedded by default, which might not be possible with the used GraphQL library).

Maybe the assets are fetched from a remote server? If so, that would introduce another potential reason not to embed it: we can't be sure systems have internet access, so this feature might not work in those cases.

I'm happy to defer this though, it's easy to change later, while we work out the exact details, I can at least see how this is useful during debugging 👍

Reply:
Had a quick look at this -- the embedded `index.html` file is 21,840 bytes.

Static assets (JS, etc) are pulled remotely on usage:

```html
  <link rel="stylesheet" href="//cdn.jsdelivr.net/npm/graphql-playground-react/build/static/css/index.css" />
  <link rel="shortcut icon" href="//cdn.jsdelivr.net/npm/graphql-playground-react/build/favicon.png" />
  <script src="//cdn.jsdelivr.net/npm/graphql-playground-react/build/static/js/middleware.js"></script>
```

There doesn't seem to be a way to disable embedding it -- [it's returned as a string here](https://github.com/async-graphql/async-graphql/blob/7efeeb543b4a9ee02f03e13cf394bf5e40d6b865/src/http/playground_source.rs) when calling `playground_source`.

This would have been better as a macro IMO, but at 21k I'm not sure it warrants too much additional effort.

I think the more import factor is allowing the user to turn it on/off. More from the standpoint of letting users decide whether they want to host a HTML server on the API port.